File: core/domain/src/main/kotlin/com/ohmyclipping/model/Organization.kt
package com.ohmyclipping.model

import java.time.Instant

/**
 * 외부 조직(경쟁사/고객사/파트너 등) 분류.
 * DB `organizations.type` CHECK 제약과 일치한다.
 */
enum class OrganizationType {
    COMPETITOR,
    CUSTOMER,
    PARTNER,
    OTHER
}

/**
 * 외부 조직 엔티티.
 *
 * Phase 3 PR2 에서 도입된 도메인. Category 와 many-to-many 로 연결하여
 * "경쟁사 A 관련 카테고리 클릭률" 같은 분석에 사용한다.
 *
 * @property tenantId 다중 테넌트 대비 placeholder. 현재는 항상 `"default"`.
 */
data class Organization(
    val id: String,
    val tenantId: String = "default",
    val name: String,
    val type: OrganizationType,
    val domain: String? = null,
    val description: String? = null,
    val createdAt: Instant = Instant.EPOCH,
    val updatedAt: Instant = Instant.EPOCH,
    /** 종목 코드 (예: 999930). V134 에서 추가. */
    val stockCode: String? = null,
    /** 조직 별칭 목록. DB 에는 JSON TEXT 로 저장. V134 에서 추가. */
    val aliases: List<String> = emptyList(),
    /** 조직 생성 경로 (user_wizard / admin_created / competitor_mirror / backfill / legacy). V134 에서 추가. */
    val origin: String? = null,
)

File: frontend/src/types/organization.ts
/**
 * Phase 3 PR2: 외부 조직 엔티티 (경쟁사/고객사/파트너 등).
 * 백엔드 `organizations.type` CHECK 제약과 일치.
 */
export type OrganizationType = "COMPETITOR" | "CUSTOMER" | "PARTNER" | "OTHER";

/** 사용자에게 노출할 type 레이블 — 드롭다운/뱃지에서 사용. */
export const ORGANIZATION_TYPE_LABELS: Record<OrganizationType, string> = {
  COMPETITOR: "경쟁사",
  CUSTOMER: "고객사",
  PARTNER: "파트너",
  OTHER: "기타",
};

/**
 * 조직 생성 경로. 백엔드 `organizations.origin` DB 값과 일치.
 * V134 신규.
 */
export type OrgOrigin =
  | "user_wizard"
  | "admin_created"
  | "competitor_mirror"
  | "backfill"
  | "legacy";

export interface Organization {
  id: string;
  tenantId: string;
  name: string;
  type: OrganizationType;
  domain: string | null;
  description: string | null;
  /** 한국 주식 종목 코드 (V134 신규). */
  stockCode: string | null;
  /** 조직 별칭 목록 (V134 신규). */
  aliases: string[];
  /** 생성 경로 (V134 신규). */
  origin: OrgOrigin | null;
  /** 연결된 카테고리 수. 목록 API 에서만 집계된다. 단건 조회 시 0. */
  usageCount: number;
  createdAt: string;
  updatedAt: string;
}

export interface OrganizationListResponse {
  content: Organization[];
  totalCount: number;
}

export interface CreateOrganizationRequest {
  name: string;
  type: OrganizationType;
  domain?: string | null;
  description?: string | null;
}

export interface UpdateOrganizationRequest {
  name?: string;
  type?: OrganizationType;
  domain?: string | null;
  description?: string | null;
  aliases?: string[] | null;
}

export interface SetCategoryOrganizationsRequest {
  organizationIds: string[];
}
